mqtt/mqtt-bridge/src/persist/mod.rs

```diff
@@ -0,0 +1,113 @@
+use std::cmp::Ordering;
+use std::time::Duration;
+
+use async_trait::async_trait;
+use futures_util::stream::Stream;
+use mqtt3::proto::Publication;
+use thiserror::Error;
+
+mod memory;
+
+// Queue used in bridge.
+#[async_trait]
+trait Persist<'a> {
+    type Loader: Stream;
+
+    fn new() -> Self;
+
+    async fn insert(
+        &mut self,
+        priority: u32,
+        ttl: Duration,
+        message: Publication,
+    ) -> Result<Key, QueueError>;
+
+    async fn remove(&mut self, key: Key) -> Result<bool, QueueError>;
+
+    async fn loader(&'a mut self, batch_size: usize) -> Self::Loader;
+}
+
+// Keys used in the queue.
+// Ordered by offset only.
+// When we implement priority and ttl for iothub telemetry we can order by priority first.
+// Unordered by ttl.
+#[derive(Eq, Ord, PartialEq, Clone, Debug)]
+pub struct Key {
+    priority: u32,
+    offset: u32,
+    ttl: Duration,
+}
+
+impl PartialOrd for Key {
+    fn partial_cmp(&self, other: &Key) -> Option<Ordering> {
+        Some(self.offset.cmp(&other.offset))
+    }
+}
+
+#[derive(Debug, Error)]
+pub enum QueueError {
+    #[error("Failed to remove messages from queue")]
+    Removal(),
+}
+
+#[cfg(test)]
+mod tests {
+    use std::time::Duration;
+
+    use crate::persist::Key;
+
+    #[test]
+    fn key_offset_ordering() {
+        // ordered by offset
+        let key1 = Key {
+            priority: 0,
+            offset: 0,
+            ttl: Duration::from_secs(5),
+        };
+        let key2 = Key {
+            priority: 0,
+            offset: 1,
+            ttl: Duration::from_secs(5),
+        };
+        assert!(key2 > key1);
+        assert!(key2 != key1);
+        assert!(key1 < key2);
+    }
+
+    #[test]
+    fn key_priority_ordering() {
+        // not ordered by priority
+        let key1 = Key {
+            priority: 0,
+            offset: 0,
+            ttl: Duration::from_secs(5),
+        };
+        let key2 = Key {
+            priority: 1,
+            offset: 0,
+            ttl: Duration::from_secs(5),
+        };
+        assert_eq!(key1 > key2, false);
+        assert_eq!(key1 < key2, false);
+        assert!(key1 != key2);
+    }
+
+    // TODO REVIEW: Does this guarantee that btreemap is not ordered by ttl? Why this weird behavior with equals?
```

Comment:
It is indeed weird behaviour here, when all key1>key2, key1<key2 and key1==key2 are all false. But it makes sense from the implementation of PartialOrd and PartialEq for Key. So, I believe, in BTreeMap the keys will be ordered by offset only (desired behaviour), but at the same time, keys with the same offset but different pri or ttl are not the same keys. It would be good to verify that with a test

Reply:
I talked to Varun and at this point we aren't thinking of removing priority / ttl in the bridge. Denis and I decided to remove it from the interface, key struct, etc.